designs/0030-complex-numbers.md

```diff
@@ -0,0 +1,355 @@
+
+The elementwise vector and matrix operations, `.*` and  `./`, should work as expected, carrying out their operations elementwise.    They should allow mixed complex and real results, e.g.,
+
+```
+vector[2] x;
+complex_vector[2] y;
+complex_vector[2] z = x .* y;
+```
+
+### Scalar functions
+
+Every scalar complex function available in the C++ standard library `<complex>` header will be included.  These are already implemented in the Stan math library.
+
+### Matrix functions
+
+There are no matrix functions exposed through the math library, but fast Fourier transforms (FFT), asymmetric eigendecomposition, and Schur decomposition have all been implemented and are part of the unit tests for complex numbers.
+
+
+## C++ model class
+
+Stan programs are converted to C++ class definitions by the stanc3 transpiler.   There are two components in this compiled class that.  This section covers data input and parameter output.  User-facing I/O is handled by the interfaces.
+
+### Data input format
+
+The C++ model class uses an object of class `var_context` to input variables declared in the data block.  This object needs to be extended to read complex number objects.  To read a complex number, we must read a sequence of two real numbers to pass to the C++ `complex<double>` constructor.
+
+### Parameter output format
+
+Parameters are output as an array.  Complex numbers will be sequenced with their real components before their imaginary components.  The output names for a complex variable `z` will be `z.r` and `z.i` for the real and imaginary components.
+
+For containers of complex numbers, the natural order is to indicate container indexing first.  For example, an entry for a matrix might look like `z.2.3.r` for the real component of the complex element at row 2 and column 3.
+
+
+## File I/O formats
+
+We have file-based output for samples and structured JSON or R dump format for input.  We use the same format for specifying data, initial values, etc., so there is only one format needed that will work for all of the interface needs.
+
+### CmdStan output
+
+CmdStan output just mirrors the C++ class output and will write headers the same way they are produced by the C++ class.
+
+### JSON input
+
+The key design aspect here is that we know when we want to read an array and know when we want to read complex numbers.  So we do not need to encode the complex/non-complex distinction directly in the output structure.  Specifically, we will follow the example in the 
+[Python JSON spec](https://docs.python.org/3/library/json.html), which encodes a complex number `3 - 1.5j` as the list `[3, -1.5]`.  
+
+A self-documenting JSON format might look like this, 
+
+```
+{ "real": 3, "imag": -1.5 }
+```
+
+but that's going to be far too verbose.
+
+### R dump format input
+
+We can adopt the standard R dump format for complex number output, which separates the real and imaginary component with the 
+
+```
+z1 <- 3-1.5i
+z2 <- 2.3+4.7i
+```
```

Comment:
This has not been implemented, but complex numbers can be read in with something like 

`z1 <- c(3, 1.5)`

similar to JSON

Reply:
OK, then this also needs to be updated.